zebra-rpc/src/sync.rs

```diff
@@ -0,0 +1,385 @@
+                    // If calling the `getblock` RPC method fails with an unexpected error, wait for the next chain tip change
+                    // without resetting the non-finalized state.
+                    Some(Err(err)) => {
+                        tracing::warn!(
+                            ?err,
+                            "encountered an unexpected error while calling getblock method"
+                        );
+
+                        break false;
+                    }
+                };
+
+                let block_hash = block.hash;
+                let commit_result = if self.non_finalized_state.chain_count() == 0 {
+                    self.non_finalized_state
+                        .commit_new_chain(block.clone(), &self.db)
+                } else {
+                    self.non_finalized_state
+                        .commit_block(block.clone(), &self.db)
+                };
+
+                // The previous block hash is checked above, if committing the block fails for some reason, try again.
+                if let Err(error) = commit_result {
+                    tracing::warn!(
+                        ?error,
+                        ?block_hash,
+                        "failed to commit block to non-finalized state"
+                    );
+
+                    break false;
+                }
+
+                while self
+                    .non_finalized_state
+                    .best_chain_len()
+                    .expect("just successfully inserted a non-finalized block above")
+                    > MAX_BLOCK_REORG_HEIGHT
+                {
+                    tracing::trace!("finalizing block past the reorg limit");
+                    self.non_finalized_state.finalize();
+                }
+
+                self.update_channels(block);
+                current_tip_hash = block_hash;
+                last_chain_tip_hash = current_tip_hash;
+
+                // If the block hash matches the output from the `getbestblockhash` RPC method, we can wait until
+                // the best block hash changes to get the next block.
+                if block_hash == target_tip_hash {
+                    break false;
+                }
+            };
+
+            if should_reset_non_finalized_state {
+                self.try_catch_up_with_primary().await;
+                let block = self.finalized_chain_tip_block().await.expect(
+                    "should have genesis block after successful bestblockheightandhash response",
+                );
+
+                last_chain_tip_hash = block.hash;
+                self.non_finalized_state =
+                    NonFinalizedState::new(&self.non_finalized_state.network);
+                self.update_channels(block);
+            }
+        }
+    }
+
+    /// Tries to catch up to the primary db instance for an up-to-date view of finalized blocks.
+    // TODO: Use getblock RPC if it fails to catch up to primary?
```

Comment:
This seems complicated and unnecessary.
```suggestion
```

Reply:
Applied in https://github.com/ZcashFoundation/zebra/pull/8596/commits/15feed173daced582340fdcdb4d4bb8cc9593a39.